### ISSUE TITLE
MPG_MODE

### ISSUE THREAD

**Issue by stevempotter** [grblHAL/Plugin_keypad#3]
I would like to use a manual pulse generator (encoder) with my keypad. Is there some documentation on how to use MPG_MODE? I have uncommented that in my_machine.h. What happens when I send 0x8B with a strobe pulse on the I2C line? Do all the other commands stop working? Is 0x8B a toggle? (enable and disable MPG alternately?)
What sorts of signals is the MPG expected to send? Are those sent via I2C or USB?
Is there a way to send a ? command to get status? Or even just the current coordinates to display on a DRO?

**Comment by terjeio** [comment]
> Is there some documentation on how to use MPG_MODE? 

I have started writing some in the [Wiki](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces).

> What happens when I send 0x8B with a strobe pulse on the I2C line?

I have not tested that, but it should work. Depending on what the controller decide it may respond with a `|MPG:1` element in a pushed real-time report that signals that the MPG now have full control via the MPG UART input stream.

> Do all the other commands stop working?

Nothing stops working, but any connected sender can no longer send commands other than the single character real-time commands (well, it can but the commands will be ignored).

> Is 0x8B a toggle? (enable and disable MPG alternately?)

Yes.

> What sorts of signals is the MPG expected to send?

When in full control it can acts as a full blown sender, if not only the mode switching command. However, if the keypad plugin is enabled it can send the single character real-time commands via I2C (KEYPAD_ENABLE 1) or via the UART input stream (KEYPAD_ENABLE 2) listed in the plugin readme.

> Are those sent via I2C or USB?

When in full control commands has to be sent via the MPG UART, when not either via I2C or via the MPG UART.

> Is there a way to send a ? command to get status?

Yes, single character real-time commands can always be sent. But if there is a sender connected requesting real-time reports the the MPG should not do it as well. It should just listen to the real-time reports requested by the sender.

> Or even just the current coordinates to display on a DRO?

No, if you want something else than a real-time report then you have to write a plugin outputting what you want.

---

FYI I have written [a parser for the real-time report](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/blob/master/Firmware/RP2040/src/grblcomm.c) and other messages from grblHAL for my MPG & DRO project that calls a callback function on changes, you may use this for your project if you want to.

**Comment by stevempotter** [comment]
Thanks, Terje. I am making progress! I have my pendant receiver (an ESP32-S2) talking to the Teensy via its RX1 and TX1 lines. 
Others will be able to get going sooner than I did if this info is easy to find in the Wiki:
Please add to the core[ MPG & DRO Interfaces Wiki](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces#keypad-plugin) that the UART referred to is not the Controller's USB port but its RX1 and TX1 GPIO pins, and that the RX1 and TX1 lines on both ends need to be 115200 baud, 8-N-1. Also, please add that all $ commands and Gcode commands have to be terminated with a CR or NL (but not both), and that the unprintable commands must have no terminators (just one byte). 

I am still trying to figure out the blocking rules of when and what I can communicate over I2C or UART when IOsender is running. The pendant is not very useful if one blocks the other. I can't get any response from the controller to my ESP32-S2 when in MPG:0 mode if I request status via UART with either ? or 0x87 or 0x80. Is there a way to query the status when in keypad mode?

I have not tried mode switching with a hardware pin. What pin would be used by default for that in the Keypad plugin? (or, how to set it to one of our unused pins?)
I am having a hard time understanding this Wiki info on Mode Switching:
"the controller pin enabled as output and driven low while it starts up, when startup is completed the pin is switched to input with pullup enabled. When the pin is driven low an app should wait for it to go high before attempting to claim control. "
Why is this line changing from output to input?

**Comment by terjeio** [comment]
I have added to the MPG & DRO interfaces wiki links to the Grbl interface and jogging wiki pages, and info about the default connection parameters. UART port and pin numbers are specifict to the driver in question and do not belong there. Which port and pins to use may also be board map specific, using the $pins command is the best way to find out?

> I am still trying to figure out the blocking rules of when and what I can communicate over I2C or UART when IOsender is running. 

In "normal" mode (`MPG:0`) the MPG may only send single character commands, the sender is in full control. Single character jogging is only allowed when the controller is in `Idle` or `Jog` mode.
In MPG mode (`MPG:1`) the ioSender UI is disabled and the MPG is in full control. A sender may still send single character commands but ioSender does not. Any other commands from a sender will be ignored.

> I have not tried mode switching with a hardware pin. What pin would be used by default for that in the Keypad plugin? (or, how to set it to one of our unused pins?)

The pin is defined [here](https://github.com/grblHAL/iMXRT1062/blob/6e91ccba8f191e17a137383612cff06a2e2d28b5/grblHAL_Teensy4/src/T41U5XBB_map.h#L68-L70). It should be possible to switch the pin from beeing an output pulled low to input with pullup during startup. FYI the T41U5XBB aux inputs does not allow that. See below.

> What pin would be used by default for that in the Keypad plugin? (or, how to set it to one of our unused pins?)

The keypad plugin does not use/repsond to the mode switch pin, it uses the `0x8B` single character command instead.
When the keypad plugin is configured to use UART input neither the I2C strobe pin or the MPG mode pin is used.

> Why is this line changing from output to input?

It is for the MPG to detect that the controller is up and can accept commands or a mode switch. As long as the line is low (during startup or not powered) it is not, the MPG should then wait for it to go high before proceeding. So at startup the MPG should set its MPG mode pin as an input \(without pullup!\) and wait for it to go high before changing it to an output and attempting a mode switch.

**Comment by stevempotter** [comment]
Thanks, that is helpful. 

> When the keypad plugin is configured to use UART input neither the I2C strobe pin or the MPG mode pin is used.

I have ganged and auto-squared axes so LIMB (M4_LIMIT_PIN) pin 28 is needed for the Y2 limit switch. Considering what pin to use for hardware mode switching, it seems that the most logical thing for me to do is to always use UART for the whole pendant (keys and MPG), never use I2C, and use the breakout board's Strobe (KPSTR) pin 41 as the hardware mode switch pin. Would that work?  Then I should use #define MPG_ENABLE 1 and #define KEYPAD_ENABLE 2 in my_machine.h, right? Or do I use #define KEYPAD_ENABLE 0? Or just comment out //#define KEYPAD_ENABLE?

For future versions of IOsender, it would be helpful 
1. to have any feedback to the MPG echoed in IOsender's console window.
2. to have IOsender's AND the MPG's DROs updated to reflect movements the MPG is commanding when in MPG:1 mode.
2. to have some way to type in single-char commands in the IOsender command entry line.

**Comment by terjeio** [comment]
>  it seems that the most logical thing for me to do is to always use UART for the whole pendant (keys and MPG), never use I2C, and use the breakout board's Strobe (KPSTR) pin 41 as the hardware mode switch pin. Would that work? Then I should use #define MPG_ENABLE 1 and #define KEYPAD_ENABLE 2 in my_machine.h, right?

I think this would require a bit of code changes in driver.c - since both MPG_ENABLE and KEYPAD_ENABLE has to be set to 2 to make them work together. And then only 0x8B is accepted for mode switching. And I do not think it will be a good idea to allow signalling a mode change with both methods at the same time. I'll have to look into this - yet another enable option to add?

> to have any feedback to the MPG echoed in IOsender's console window.

This is a tricky one, will all senders accept unsolicited reports without issues? I have been pondering about adding a command to grblHAL where the sender identifies itself (or rather its capabilities) for this and other purposes as well. I do push the real-time report when possible, see below, but not any of the other reports output by $$, $G etc.

> to have IOsender's AND the MPG's DROs updated to reflect movements the MPG is commanding when in MPG:1 mode.

Another tricky one that I have not yet been able to solve when the communication channel to the sender is over native USB. The issue is to be sure that there is a client (sender) connected that will read data - if not the USB stream will block when the output buffer is full. If the sender connection is via another UART stream or a network stream the sender UI should update, A UART connection because data is transmitted regardless of there beeing a client at the other end and network streams since they provide connect/disconnect events.

> to have some way to type in single-char commands in the IOsender command entry line.

Have you tried the OS provided ALT key sequences already? IIRC it is ALT 0 followed by three numbers for the decimal value of the character.

**Comment by stevempotter** [comment]
> >to have IOsender's AND the MPG's DROs updated to reflect movements the MPG is commanding when in MPG:1 mode.
> 
> Another tricky one that I have not yet been able to solve when the communication channel to the sender is over native USB.

I am using the ethernet interface of Teensy to talk to IOsender, which is working great. Does that make it easier to keep two DROs synced during jogging?

I will give the ALT 0 thing a try. I did not know about that (I usually use a Mac for other things than CNC).

**Comment by terjeio** [comment]
> I am using the ethernet interface of Teensy to talk to IOsender, which is working great. Does that make it easier to keep two DROs synced during jogging?

Yes, that should work out-of-the-box unless I have made a mistake when tinkering with the code. I have to reverify and plan to do that, my setup for MPG testing is currently wired to a Pi Pico controller so it may take a little while.

BTW I have committed an update that should allow using the mode change pin even when the keypad plugin is enabled:
```
//#define MPG_ENABLE          1 // Enable MPG interface. Requires serial port and one handshake pin unless
                                // KEYPAD_ENABLE is set to 2 when mode switching is done by the CMD_MPG_MODE_TOGGLE (0x8B)
                                // command character. Set both MPG_ENABLE and KEYPAD_ENABLE to 2 to use a handshake pin anyway.
//#define KEYPAD_ENABLE       1 // Set to 1 for I2C keypad, 2 for other input such as serial data. If KEYPAD_ENABLE is set to 2
                                // and MPG_ENABLE is uncommented then a serial stream is shared with the MPG.
```

**Comment by stevempotter** [comment]
Great - thanks for the updates. I flashed in the new grblHAL with MPG_Enable 2 and KEYPAD_ENABLE 2, and both seem to be working. I have successful hardware switching between MPG:0 and MPG:1 using pin 41 of the Teensy. I can execute long commands from pendant receiver via UART in MPG:1 and keypad single letter commands via UART in MPG:0. 

IOsender is echoing its periodic status reports to UART in MPG:0, but not other things appearing in the IOsender Console window, such as responses to $$ or $help. It would be nice to echo those, too, or at least have that option.

When in MPG:1 (and IOsender is greyed out), the DRO of IOsender is updated only when my pendant asks for a status update with 0x87 or 0x80, which is fine.

Maybe the following is an IOsender problem?: Running a test code (the grblHAL logo engraving), I don't see anything in the Feed rate box while running the program. It stays at zero. And the Feed rate override buttons don't seem to work.

Re: single char commands:

>>to have some way to type in single-char commands in the IOsender command entry line.

>Have you tried the OS provided ALT key sequences already? IIRC it is ALT 0 followed by three numbers for the decimal value of the character.

Typing any character in IOsender input box with the ALT key pressed causes a system error chime. So it seems that the ALT+0xxx commands are filtered out. Can you allow them?

**Comment by stevempotter** [comment]
I forgot to mention that I also added to T41U5XBB_map.h
#define MPG_MODE_PIN (41u)
to enable the hardware mode switch.

**Comment by terjeio** [comment]
I have added keyboard shortcuts for some single character commands in the [latest edge versions](https://github.com/terjeio/ioSender/discussions/176#discussioncomment-2220757).
```
ALT 0 1 3 2 -> CMD_SAFETY_DOOR (0x84)
ALT 0 1 3 5 -> CMD_STATUS_REPORT_ALL (0x87)
ALT 0 1 3 6 -> CMD_OPTIONAL_STOP_TOGGLE (0x88)
ALT 0 1 3 7 -> CMD_SINGLE_BLOCK_TOGGLE (0x89)
ALT 0 1 6 4 -> CMD_PROBE_CONNECTED_TOGGLE (0x89)
```
The remaining relevant codes are supported by checkboxes/buttons in the UI so see no reason for duplicating them. I had to translate them back to 8-bit ASCII as the most of the keycodes were translated by Windows from ASCII to UTF-8.

Note that the numeric input has to be via the keys in the numeric part of keyboard (not the keys above the letter keys).

**Comment by AvatarBg111** [comment]
Hello. I'm building my own 3-axis CNC machine and i have the following ideas:
 - I'm thinking of having two microcontrollers in the machine - one for controlloing the machine (a STM32F407VGT6) and another one for controlling the control panel of the machine (a STM32H750VBT6).
 - I want to be able to control the machine by sending it g-code through USB (directly to the stm32f4) and have manual control by sending commands and g-code between the two STM's through a UART interface.
I analyzed a little bit of the grblHAL firmware and came to the conclusion that commands and g-code can be received by the microcontroller (the one with th grblHAL firmware) through 3 interfaces - 2 UARTs and 1 USB. One of the UARTs is dedicated for receiving g-code, while the other one is for MPG (at least that's what i understood from the software).

Now here's my problem - is there a way for me to switch between USB and MPG UART without having to reflash the microcontroller (STM32F407)? Or can I switch between the USB serial and the dedicated UART serial?

**Comment by terjeio** [comment]
> Now here's my problem - is there a way for me to switch between USB and MPG UART without having to reflash the microcontroller (STM32F407)? Or can I switch between the USB serial and the dedicated UART serial?

See the Wiki [here](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces#mpg--dro-mode).

**Comment by AvatarBg111** [comment]
> > Now here's my problem - is there a way for me to switch between USB and MPG UART without having to reflash the microcontroller (STM32F407)? Or can I switch between the USB serial and the dedicated UART serial?
> 
> See the Wiki [here](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces#mpg--dro-mode).

I tried different configurations of the firmware by changing the values of KEYPAD_ENABLE and MPG_ENABLE while trying to keep the USB as the main serial communication method, but I found out that if MPG_ENABLE is not defined and KEYPAD_ENABLE is - the USB works and the app on my computer (LaserGRBL) can connect to my microcontroller, but if MPG_ENABLE is defined, a connection through USB cannot be made with a PC. Why is that? Is it even possible to have a a machine, which receives g-code from USB, but can also be manually controlled with jogging commands from a "pendant"/control panel, without changing the code of the firmware? Can this problem be solved if MPG_ENABLE is not defined and only KEYPAD_ENABLE is and the jogging commands can be sent to the dedicated keypad plugin UART/I2C peripheral?
